likelihood/cl_like/bacco.py

```diff
@@ -0,0 +1,335 @@
+            'ns': cosmo['n_s'],
+            #TODO: Remove massive neutrinos from sigma_8
+            'sigma8': cosmo['sigma8'],
+            'neutrino_mass': np.sum(cosmo['m_nu']),
+            'w0': cosmo['w0'],
+            'wa': cosmo['wa'],
+            'expfactor': 1.  # random scale factor just to initialize
+        }
+        return pars
+
+    def update_pk(self, cosmo):
+        """ Update the internal PT arrays.
+
+        Args:
+            pk (array_like): linear power spectrum sampled at the
+                internal `k` values used by this calculator.
+        """
+
+        # translate the pyccl cosmology parameters into bacco notation
+        pars = self._cosmo_to_bacco(cosmo)
+        #TODO: Do this only once?
+        lpt_emu = baccoemu.Lbias_expansion()
+        # convert k_s [Mpc^-1] into h Mpc^-1 units just for the calculation
+        k = self.ks / pars['hubble']
+
+        # 10 redshifts, 15 combinations between bias params, and the ks (added b0)
+        num_comb = 15
+        pk2d_bacco = np.zeros((len(self.a_s), num_comb, len(self.ks)))
+
+        # compute the power for each redshift
+        # t1 = time.time()
+        for i in range(len(self.a_s)):
+            pars['expfactor'] = self.a_s[i]
+            # call the emulator of the nonlinear 15 lagrangian bias expansion terms, shape is (15, len(k))
+            k, pnn = lpt_emu.get_nonlinear_pnn(pars, k=k)
+            pk2d_bacco[i, :, :] = pnn
+
+        # convert the spit out result from (Mpc/h)^3 to Mpc^3 (bacco uses h units, but pyccl doesn't)
+        pk2d_bacco /= pars['hubble'] ** 3
+        self.bacco_table = pk2d_bacco
+
+    def get_pgg(self, Pnl, b11, b21, bs1, b12, b22, bs2,
+                bk21=None, bk22=None, Pgrad=None):
+        if self.bacco_table is None:
+            raise ValueError("Please initialise BACCO calculator")
+        # Clarification:
+        # BACCO uses the following expansion for the galaxy overdensity:
+        #   d_g = b1 d + b2 d2^2 + bs s^2 + bnabla2d nabla^2 d
+        # (see Eq. 1 of https://arxiv.org/abs/2101.12187).
+        #
+        # The BACCO table below contains the following power spectra
+        # in order:
+        # <1,1>
+        # <1,d>
+        # <1,d^2>
+        # <1,s^2>
+        # <1,nabla^2 d>
+        # <d,d>
+        # <d,d^2>
+        # <d,s^2>
+        # <d,nabla^2 d>
+        # <d^2,d^2> (!)
+        # <d^2,s^2> (!)
+        # <d^2,nabla^2 d>
+        # <s^2,s^2> (!)
+        # <s^2,nabla^2 d>
+        # <nabla^2 d,nabla^2 d>
+        #
+        # So:
+        #   a) The spectra involving b2 are for d^2 - convert to d^2/2
+        #   b) The spectra involving bs are for s^2 - convert to s^2/2
+        # Also, the spectra marked with (!) tend to a constant
+        # as k-> 0, which we can suppress with a low-pass filter.
+        #
+        # Importantly, we have corrected the spectra involving s2 to
+        # make the definition of bs equivalent in the EPT and LPT
+        # expansions.
```

Comment:
Is this part of the comment just a leftover from a similar comment in `lpt.py`?

Reply:
Thanks so so much for going though the code and all your comments @damonge! 
Thanks a lot for this! Yes totally - I fixed it in the latest commit.